### PR TITLE
Llu/softmax

### LINKED ISSUE #2125: Bad performance in a softmax fusion
### 🐛 Describe the bug

Repro:

```
import torch
from torch._C._nvfuser import FusionDefinition, Fusion, DataType

def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, 1, 1, -1], contiguous=[True, True, True, True], dtype=DataType.Float)
    T1 = fd.define_tensor(symbolic_sizes=[-1, -1, -1, -1], contiguous=[True, True, True, True], dtype=DataType.Float)
    S2 = fd.define_constant(1)
    T3 = fd.ops.mul(T0, S2)
    S4 = fd.define_constant(8.00000)
    T5 = fd.ops.div(T1, S4)
    S6 = fd.define_constant(1.00000)
    T7 = fd.ops.sub(S6, T3)
    S8 = fd.define_constant(-10000.0)
    T9 = fd.ops.mul(T7, S8)
    T10 = fd.ops.broadcast_in_dim(T9, output_shape=[2, 64, 512, 512], broadcast_dims=[0, 1, 2, 3])
    T11 = fd.ops.add(T5, T10)
    T12 = fd.ops.max(T11, axes=[3], keepdim=False, dtype=DataType.Null)
    T13 = fd.ops.broadcast_in_dim(T12, output_shape=[2, 64, 512, 1], broadcast_dims=[0, 1, 2])
    T14 = fd.ops.broadcast_in_dim(T13, output_shape=[2, 64, 512, 512], broadcast_dims=[0, 1, 2, 3])
    T15 = fd.ops.sub(T11, T14)
    T16 = fd.ops.exp(T15)
    T17 = fd.ops.sum(T16, axes=[3], keepdim=False, dtype=DataType.Null)
    T18 = fd.ops.broadcast_in_dim(T17, output_shape=[2, 64, 512, 1], broadcast_dims=[0, 1, 2])
    T19 = fd.ops.broadcast_in_dim(T18, output_shape=[2, 64, 512, 512], broadcast_dims=[0, 1, 2, 3])
    T20 = fd.ops.div(T16, T19)
    fd.add_output(T9)
    fd.add_output(T20)

fs = Fusion()
with FusionDefinition(fs) as fd:
    nvfuser_fusion_id1(fd)

inputs = [
    torch.randn(2, 1, 1, 512, device='cuda'),
    torch.randn(2, 64, 512, 512, device='cuda'),
]

for _ in range(20):
    out = fs.execute(inputs)
```

Relevant debug:

```
 PYTORCH_NVFUSER_DUMP=dump_eff_bandwidth,ptxas_verbose,launch_param,scheduler_params
 ```
 Output:
 
 ```
 ptxas info    : 3 bytes gmem
ptxas info    : Compiling entry function '_ZN11CudaCodeGen7kernel1ENS_6TensorIfLi4EEES1_S1_S1_' for 'sm_80'
ptxas info    : Function properties for _ZN11CudaCodeGen7kernel1ENS_6TensorIfLi4EEES1_S1_S1_
ptxas         .     0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 26 registers, 16 bytes smem, 512 bytes cmem[0], 8 bytes cmem[2]

Launch Parameters: BlockDim.x = 128, BlockDim.y = 4, BlockDim.z = -1, GridDim.x = 16384, GridDim.y = -1, GridDim.z = -1, Smem Size = 2048
kernel1 run in 0.970752 ms, achieved: 276.532 GB/s
Launch Parameters: BlockDim.x = 128, BlockDim.y = 4, BlockDim.z = -1, GridDim.x = 16384, GridDim.y = -1, GridDim.z = -1, Smem Size = 2048
kernel1 run in 0.907264 ms, achieved: 295.883 GB/s

===== Reduction Stats ========
total_reduction_numel: 512
total_iteration_numel: 65536
inner_most_dimension_numel: 512
vectorize_factor: 4
n_tensor_inputs: 2
max_input_dtype_size: 4
max_persistent_buffer_size: 2048
max_multi_reduction_factor: 64
block(128, 4, 1)
===== Reduction Parameters ========
Tag: Inner Persistent Heuristic.

Red On Fastest Dim
Persistent Kernel
Batches per block: 1

Iteration Domain: blockIdx.x / threadIdx.y / multiple reductions per block / 
Inner Reduction Domain: cross block - threadIdx.x / persistent batch - 1 / vectorize / factor 4
Launch Parameters: BlockDim.x = -1, BlockDim.y = 4, BlockDim.z = -1, GridDim.x = -1, GridDim.y = -1, GridDim.z = -1, Smem Size = 0
```
FX definition:

```
class fused_49(torch.nn.Module):                                                                                                                                                                                                                                                                                                                                            
    def forward(self, broadcast_in_dim_2: f32[2,1,1,512], view_13: f32[2,64,512,512]):                                                                                                                                                                                                                                                                                      
        # No stacktrace found for following nodes                                                                                                                                                                                                                                                                                                                           
        mul: f32[2,1,1,512] = torch.ops.nvprims.mul.default(broadcast_in_dim_2, 1);  broadcast_in_dim_2 = None                                                                                                                                                                                                                                                              
        div: f32[2,64,512,512] = torch.ops.nvprims.div.default(view_13, 8.0);  view_13 = None                                                                                                                                                                                                                                                                               
        sub: f32[2,1,1,512] = torch.ops.nvprims.sub.default(1.0, mul);  mul = None                                                                                                                                                                                                                                                                                          
        mul_1: f32[2,1,1,512] = torch.ops.nvprims.mul.default(sub, -10000.0);  sub = None                                                                                                                                                                                                                                                                                   
        broadcast_in_dim_12: f32[2,64,512,512] = torch.ops.nvprims.broadcast_in_dim.default(mul_1, [2, 64, 512, 512], [0, 1, 2, 3])                                                                                                                                                                                                                                         
        add_4: f32[2,64,512,512] = torch.ops.nvprims.add.default(div, broadcast_in_dim_12);  div = broadcast_in_dim_12 = None                                                                                                                                                                                                                                               
        amax: f32[2,64,512] = torch.ops.nvprims.amax.default(add_4, [3])                                                                                                                                                                                                                                                                                                    
        broadcast_in_dim_13: f32[2,64,512,1] = torch.ops.nvprims.broadcast_in_dim.default(amax, [2, 64, 512, 1], [0, 1, 2]);  amax = None                                                                                                                                                                                                                                   
        broadcast_in_dim_14: f32[2,64,512,512] = torch.ops.nvprims.broadcast_in_dim.default(broadcast_in_dim_13, [2, 64, 512, 512], [0, 1, 2, 3]);  broadcast_in_dim_13 = None                                                                                                                                                                                              
        sub_2: f32[2,64,512,512] = torch.ops.nvprims.sub.default(add_4, broadcast_in_dim_14);  add_4 = broadcast_in_dim_14 = None                                                                                                                                                                                                                                           
        exp: f32[2,64,512,512] = torch.ops.nvprims.exp.default(sub_2);  sub_2 = None                                                                                                                                                                                                                                                                                        
        sum_1: f32[2,64,512] = torch.ops.nvprims.sum.default(exp, [3])                                                                                                                                                                                                                                                                                                      
        broadcast_in_dim_15: f32[2,64,512,1] = torch.ops.nvprims.broadcast_in_dim.default(sum_1, [2, 64, 512, 1], [0, 1, 2]);  sum_1 = None                                                                                                                                                                                                                                 
        broadcast_in_dim_16: f32[2,64,512,512] = torch.ops.nvprims.broadcast_in_dim.default(broadcast_in_dim_15, [2, 64, 512, 512], [0, 1, 2, 3]);  broadcast_in_dim_15 = None                                                                                                                                                                                              
        div_1: f32[2,64,512,512] = torch.ops.nvprims.div.default(exp, broadcast_in_dim_16);  exp = broadcast_in_dim_16 = None                                                                                                                                                                                                                                               
        return (mul_1, div_1)
```
Relevant Benchmark command:

```
python -u benchmarks/huggingface.py --training -d cuda --fast --backend nvprims_nvfuser --skip-accuracy-check --performance --only AlbertForMaskedLM
```

Cuda Kernel:

```
======= Codegen output for kernel: kernel1 =======                                                                                                                                                                                                                                                                                                                          
                                                                                           
__global__ void kernel1(Tensor<float, 4> T0, Tensor<float, 4> T1, Tensor<float, 4> T22, Tensor<float, 4> T19) {                                                                       
  alignas(16) extern __shared__ char array[];                                                                                                                                         
  void* shared_mem = array;                                                                                                                                                           
  NVFUSER_DEFINE_MAGIC_ZERO
  double d9;                
  d9 = (double)(1);                                                                                                                                                                                                                                                                                                                                                         
  int i344;                                                                                                                                                                                                                                                                                                                                                                 
  i344 = ((((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.y)) + ((nvfuser_index_t)threadIdx.y)) / (T1.size[1] * T1.size[2]);                                              
  int i337;                                                           
  i337 = (((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.y)) + ((nvfuser_index_t)threadIdx.y);                                                                            
  int i335;                                                                                                                                                                           
  i335 = ((nvfuser_index_t)threadIdx.x) * 4;                                         
  float T8[((1 * 4) * 1)]; 
  if ((((i337 < (T0.size[0] * (T1.size[1] * T1.size[2]))) && (((((nvfuser_index_t)threadIdx.x) * 4) + 3) < T0.size[3])) && (i344 < T0.size[0]))) {                                    
    Array<float, (4 * 1), 4> T21;                                     
    T21.set(0);            
    loadGlobalToLocal<float, 4, false>(&T21[0],  &T1[(i337 * T0.size[3]) + i335]);
    Array<float, (4 * 1), 4> T20;                                     
    T20.set(0);                                                                                                                                                                       
    loadGlobalToLocal<float, 4, false>(&T20[0],  &T0[(i344 * T0.size[3]) + i335]);
    // Alias Allocation - register                                    
    auto& T23 = T20;                            
    #pragma unroll                                                    
    for(nvfuser_index_t i301 = 0; i301 < 4; ++i301) {                                                                                                                                 
      float T2[1];                  
      T2[0]                  
        = T20[i301]                                                   
        * (float) d9;       
      float T4[1];                                   
      T4[0]                                                                                                                                                                           
        = (float) 1.00000000000000000e+00                 
        - T2[0];                                                                                                                                                                      
      float T5[1];                                                                                                                                                                    
      T5[0]                                                                                                                                                                           
        = T4[0]                                                
        * (float) -1.00000000000000000e+04;
      T23[i301]   
         = T5[0];
      float T3[1];                                                                                                                                                                                                                                                                                                                                                          
      T3[0]                                                                                                                                                                                                                                                                                                                                                                 
        = T21[i301]                                                                                                                                                                                                                                                                                                                                                         
        / (float) 8.00000000000000000e+00;                                                                                                                                                                                                                                                                                                                                  
      float T6[1];                                                                                                                                                                                                                                                                                                                                                          
      T6[0]                                                                                                                                                                                                                                                                                                                                                                 
         = T5[0];                                                                                                                                                                                                                                                                                                                                                           
      float T7[1];                                                                                                                                                                                                                                                                                                                                                          
      if (((i344 < T0.size[0]) && (((((nvfuser_index_t)threadIdx.x) * 4) + (i301 + nvfuser_zero)) < T0.size[3]))) {                                                                                                                                                                                                                                                         
        T7[0]                                                                                                                                                                                                                                                                                                                                                               
           = T6[0];                                                                                                                                                                                                                                                                                                                                                         
      }                                                                                                                                                                                                                                                                                                                                                                     
      T8[i301]                                                                                                                                                                                                                                                                                                                                                              
        = T3[0]                                                                                                                                                                                                                                                                                                                                                             
        + T7[0];                                                                                                                                                                                                                                                                                                                                                            
    }                                                                                                                                                                                                                                                                                                                                                                       
    NVFUSER_UPDATE_MAGIC_ZERO                                                                                                                                                                                                                                                                                                                                               
    if (((i344 < T0.size[0]) && (((((nvfuser_index_t)threadIdx.x) * 4) + 3) < T0.size[3]))) {                                                                                                                                                                                                                                                                               
      loadLocalToGlobal<float, 4, false>( &T22[(i344 * T0.size[3]) + i335], &T23[0]);                                                                                                                                                                                                                                                                                       
    }                                                                                                                                                                                                                                                                                                                                                                       
  }                                                                                                                                                                                                                                                                                                                                                                         
  float T9[1];                                                                                                                                                                                                                                                                                                                                                              
  T9[0] = NEG_INFINITY;                                                                                                                                                                                                                                                                                                                                                     
  float T25[1];                                                                                                                                                                                                                                                                                                                                                             
  T25[0] = NEG_INFINITY;                                                                                                                                                                                                                                                                                                                                                    
  if (((i337 < (T0.size[0] * (T1.size[1] * T1.size[2]))) && (((((nvfuser_index_t)threadIdx.x) * 4) + 3) < T0.size[3]))) {                                                                                                                                                                                                                                                   
    #pragma unroll                                                                                                                                                                                                                                                                                                                                                          
    for(nvfuser_index_t i304 = 0; i304 < 4; ++i304) {                                                                                                                                                                                                                                                                                                                       
      T25[0] = fmax(                                                                                                                                                                                                                                                                                                                                                        
        T25[0],                                                                                                                                                                                                                                                                                                                                                             
        T8[i304]);                                                                                                                                                                                                                                                                                                                                                          
    }                                                                                                                                                                                                                                                                                                                                                                       
    NVFUSER_UPDATE_MAGIC_ZERO                                                                                                                                                                                                                                                                                                                                               
  } else {                                                                                                                                                                                                                                                                                                                                                                  
    #pragma unroll                                                                                                                                                                                                                                                                                                                                                          
    for(nvfuser_index_t i304 = 0; i304 < 4; ++i304) {                                                                                                                                                                                                                                                                                                                       
      if (((i337 < (T0.size[0] * (T1.size[1] * T1.size[2]))) && (((((nvfuser_index_t)threadIdx.x) * 4) + (i304 + nvfuser_zero)) < T0.size[3]))) {                                                                                                                                                                                                                           
        T25[0] = fmax(                                                                                                                                                                                                                                                                                                                                                      
          T25[0],                                                                                                                                                                                                                                                                                                                                                           
          T8[i304]);                                                                                                                                                                                                                                                                                                                                                        
      }                                                                                                                                                                                                                                                                                                                                                                     
    }                                                                                                                                                                                                                                                                                                                                                                       
    NVFUSER_UPDATE_MAGIC_ZERO                                                                                                                                                                                                                                                                                                                                               
  }                                                                                                                                                                                                                                                                                                                                                                         
  blockReduce<true, false, false>(
    T9[0],
    T25[0],                                                                                                                                                                           
    [](float &a, float b) { a = fmax(a, b); },
    threadIdx,    
    blockDim,                                        
    static_cast<float*>(shared_mem),
    true,          
    true,        
    float(NEG_INFINITY));
  float T10[1];              
  broadcast::blockBroadcast<true, false, false>(                                   
    T10[0],
    T9[0],                       
    static_cast<float*>(shared_mem),
    true);                                           
  float T11[1];
  T11[0]           
     = T10[0];   
  float T12[1];
  if ((i337 < (T0.size[0] * (T1.size[1] * T1.size[2])))) {
    T12[0]                                                                                                                                                                            
       = T11[0];                                                                     
  } else {
    if ((i337 < (T0.size[0] * (T1.size[1] * T1.size[2])))) {
      T12[0]
         = T11[0];
    }                                 
  }
  float T14[((1 * 4) * 1)];                                                                                                                                                                                                                                                                                                                                        [60/1984]
  #pragma unroll                                                                                                                                                                                                                                                                                                                                                            
  for(nvfuser_index_t i308 = 0; i308 < 4; ++i308) {                                                                                                                                                                                                                                                                                                                         
    float T13[1];                                                                                                                                                                                                                                                                                                                                                           
    T13[0]                                                                                                                                                                                                                                                                                                                                                                  
      = T8[i308]                                                                                                                                                                                                                                                                                                                                                            
      - T12[0];                                                                                                                                                                                                                                                                                                                                                             
    T14[i308]                                                                                                                                                                                                                                                                                                                                                               
       = expf(T13[0]);                                                                                                                                                                                                                                                                                                                                                      
  }                                                                                                                                                                                                                                                                                                                                                                         
  NVFUSER_UPDATE_MAGIC_ZERO                                                                                                                                                                                                                                                                                                                                                 
  float T15[1];                                                                                                                                                                                                                                                                                                                                                             
  T15[0] = 0.00000000000000000e+00;                                                                                                                                                                                                                                                                                                                                         
  float T26[1];                                                                                                                                                                                                                                                                                                                                                             
  T26[0] = 0.00000000000000000e+00;                                                                                                                                                                                                                                                                                                                                         
  if (((i337 < (T0.size[0] * (T1.size[1] * T1.size[2]))) && (((((nvfuser_index_t)threadIdx.x) * 4) + 3) < T0.size[3]))) {                                                                                                                                                                                                                                                   
    #pragma unroll                                                                                                                                                                                                                                                                                                                                                          
    for(nvfuser_index_t i311 = 0; i311 < 4; ++i311) {                                                                                                                                                                                                                                                                                                                       
      T26[0]                                                                                                                                                                                                                                                                                                                                                                
        = T26[0]                                                                                                                                                                                                                                                                                                                                                            
        + T14[i311];                                                                                                                                                                                                                                                                                                                                                        
    }                                                                                                                                                                                                                                                                                                                                                                       
    NVFUSER_UPDATE_MAGIC_ZERO                                                                                                                                                                                                                                                                                                                                               
  } else {                                                                                                                                                                                                                                                                                                                                                                  
    #pragma unroll                                                                                                                                                                                                                                                                                                                                                          
    for(nvfuser_index_t i311 = 0; i311 < 4; ++i311) {                                                                                                                                                                                                                                                                                                                       
      if (((i337 < (T0.size[0] * (T1.size[1] * T1.size[2]))) && (((((nvfuser_index_t)threadIdx.x) * 4) + (i311 + nvfuser_zero)) < T0.size[3]))) {                                                                                                                                                                                                                           
        T26[0]                                                                                                                                                                                                                                                                                                                                                              
          = T26[0]                                                                                                                                                                                                                                                                                                                                                          
          + T14[i311];                                                                                                                                                                                                                                                                                                                                                      
      }                                                                                                                                                                                                                                                                                                                                                                     
    }                                                                                                                                                                                                                                                                                                                                                                       
    NVFUSER_UPDATE_MAGIC_ZERO                                                                                                                                                                                                                                                                                                                                               
  }                                                                                                                                                                                                                                                                                                                                                                         
  blockReduce<true, false, false>(                                                                                                                                                                                                                                                                                                                                          
    T15[0],                                                                                                                                                                                                                                                                                                                                                                 
    T26[0],                                                                                                                                                                                                                                                                                                                                                                 
    [](float &a, float b) { a = a + b; },                                                                                                                                                                                                                                                                                                                                   
    threadIdx,                                                                                                                                                                                                                                                                                                                                                              
    blockDim,                                                                                                                                                                                                                                                                                                                                                               
    static_cast<float*>(shared_mem),                                                                                                                                                                                                                                                                                                                                        
    true,                                                                                                                                                                                                                                                                                                                                                                   
    true,                                                                                                                                                                                                                                                                                                                                                                   
    float(0.00000000000000000e+00));                                                                                                                                                                                                                                                                                                                                        
  float T16[1];                                                                                                                                                                                                                                                                                                                                                             
  broadcast::blockBroadcast<true, false, false>(                                                                                                                                                                                                                                                                                                                            
    T16[0],                                                                                                                                                                                                                                                                                                                                                                 
    T15[0],                                                                                                                                                                                                                                                                                                                                                                 
    static_cast<float*>(shared_mem),                                                                                                                                                                                                                                                                                                                                        
    true);                                                                                                                                                                                                                                                                                                                                                                  
  float T17[1];                                                                                                                                                                                                                                                                                                                                                             
  T17[0]                                                                                                                                                                                                                                                                                                                                                                    
     = T16[0];                                                                                                                                                                                                                                                                                                                                                              
  float T18[1];
  if ((i337 < (T0.size[0] * (T1.size[1] * T1.size[2])))) {
    T18[0]
       = T17[0];
  } else {
    if ((i337 < (T0.size[0] * (T1.size[1] * T1.size[2])))) {
      T18[0]
         = T17[0];
    }
  }
  if (((i337 < (T0.size[0] * (T1.size[1] * T1.size[2]))) && (((((nvfuser_index_t)threadIdx.x) * 4) + 3) < T0.size[3]))) {
    Array<float, (4 * 1), 4> T24;
    #pragma unroll
    for(nvfuser_index_t i313 = 0; i313 < 4; ++i313) {
      T24[i313]
        = T14[i313]
        / T18[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    loadLocalToGlobal<float, 4, false>( &T19[(i337 * T0.size[3]) + i335], &T24[0]);
  } else {
    Array<float, (4 * 1), 4> T24;
    #pragma unroll
    for(nvfuser_index_t i313 = 0; i313 < 4; ++i313) {
      T24[i313]
        = T14[i313]
        / T18[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    if (((i337 < (T0.size[0] * (T1.size[1] * T1.size[2]))) && (((((nvfuser_index_t)threadIdx.x) * 4) + 3) < T0.size[3]))) {
      loadLocalToGlobal<float, 4, false>( &T19[(i337 * T0.size[3]) + i335], &T24[0]);
    }
  }
}

======================================
```



### Versions

`csarofeen/torchbenchPerf`

### PR DESCRIPTION
**Pull request by liqiangxl**
Fixes #2125

### CODE REVIEW

File: third_party/nvfuser/csrc/ir_internal_nodes.h [hunk 1]
```diff
@@ -1394,6 +1395,7 @@ class TORCH_CUDA_CU_API IterDomainBuilder {
   bool is_padded_dimension_ = false;
   c10::optional<int64_t> padded_to_size_ = c10::nullopt;
   bool is_mma_swizzled_ = false;
+  bool is_merged_from_broadcast_ = false;
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/ir_internal_nodes.h:1398]: It looks like this is only used in lowering. It should probably be added to the concretized broadcast information, not as a member to IterDomain.

**Review by csarofeen — CHANGES_REQUESTED** [review]

**Comment by csarofeen** [comment]
what's the perf delta for the softmax case?

Sorry didn't meant to close the PR, reopened it.

File: third_party/nvfuser/csrc/ir_internal_nodes.h (hunk 1 above)
Comment (liqiangxl) [reply third_party/nvfuser/csrc/ir_internal_nodes.h:1398]: > what's the perf delta for the softmax case?
> 
For the case in https://github.com/csarofeen/pytorch/issues/2125
Before this fix was 340~370 GB/s on A100-80G, after this fix increased to 1.19e+03 GB/s on the same GPU.

Comment (csarofeen) [reply third_party/nvfuser/csrc/ir_internal_nodes.h:1398]: Very nice.

**Comment by liqiangxl** [comment]
move to https://github.com/NVIDIA/Fuser/pull/11

**Comment by liqiangxl** [comment]
moved to https://github.com/NVIDIA/Fuser/pull/11

File: third_party/nvfuser/csrc/ir_internal_nodes.h (hunk 1 above)
Comment (liqiangxl) [reply third_party/nvfuser/csrc/ir_internal_nodes.h:1398]: > It looks like this is only used in lowering. It should probably be added to the concretized broadcast information, not as a member to IterDomain.

This parameter is deleted in the revised version. The infomation can be reconstructed by itering over exprs between leaf domains and root domain.
```
all_exp = DependencyCheck::getAllExprsBetween(
        {from.begin(), from.end()},
        {out_tv->domain()->domain().begin(), out_tv->domain()->domain().end()});
```

**Pull request closed without merging**